File: server/src/main/java/net/malevy/mazeserver/domain/Directions.java
package net.malevy.mazeserver.domain;

import java.util.EnumSet;

public enum Directions {
    NORTH,
    SOUTH,
    EAST,
    WEST,
    EXIT;

    /**
     * A set of all directions
     */
    public static final EnumSet<Directions> ALL() {
        return EnumSet.of(NORTH, SOUTH, EAST, WEST);
    }

    ;

    /**
     * Factory method for creating an empty set of Directions
     */
    public static final EnumSet<Directions> EMPTY() {
        return EnumSet.noneOf(Directions.class);
    }

    ;

    /**
     * Returns the direction opposite the given direction
     *
     * @param d the given direction
     * @return the opposite direction
     * @throws IndexOutOfBoundsException when the given direction is not mapped for an opposite
     */
    public static Directions oppositeOf(Directions d) {
        if (d == NORTH) return SOUTH;
        else if (d == SOUTH) return NORTH;
        else if (d == EAST) return WEST;
        else if (d == WEST) return EAST;
        else throw new IndexOutOfBoundsException("Unmapped direction: " + d);
    }

}

File: react-client/src/services/Directions.js
const directions = ["north", "east", "south", "west"];
const directionsLen = directions.length;
const oppositeOffset = 2;

const funcs = {
  /**
   * getDirectionsComingFrom
   * @param {string} comingFrom - the origination point
   * @returns {string[]} - array of directions as they appear on a compass if observed coming
   *                          from the given direction
   */
  getDirectionsComingFrom(comingFrom) {
    const start = directions.findIndex((d) => d === comingFrom);
    let results = [];
    for (let offset = -1; offset <= 1; offset++) {
      results.push(
        directions[
          (((start + offset) % directionsLen) + directionsLen) % directionsLen
        ]
      );
    }
    return results;
  },

  /**
   * oppositeOf
   * @param {string} direction
   * @returns {string} the compass direction that is opposite of the given direction.
   * @example oppositeOf("north") => "south"
   */
  oppositeOf(direction) {
    const index = directions.findIndex((d) => d === direction);
    return directions[
      (((index + oppositeOffset) % directionsLen) + directionsLen) %
        directionsLen
    ];
  },

  EXIT: "EXIT",
};

export default funcs;
